service/feeder/live.go

```diff
@@ -0,0 +1,140 @@
+package feeder
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+
+	"github.com/fxamacker/cbor/v2"
+	"github.com/klauspost/compress/zstd"
+	"github.com/optakt/flow-dps/models/dps"
+	"github.com/rs/zerolog"
+	"go.nanomsg.org/mangos/v3"
+	"go.nanomsg.org/mangos/v3/protocol/sub"
+
+	// register transports
+	_ "go.nanomsg.org/mangos/v3/transport/all"
+
+	"github.com/onflow/flow-go/ledger"
+)
+
+const (
+	pubSubPort = 14532
+	reqPort    = 14533
+)
+
+type Live struct {
+	log zerolog.Logger
+
+	pubsub mangos.Socket
+	req    mangos.Socket
+
+	updatesMu    sync.Mutex
+	updates      []ledger.TrieUpdate
+	decompressor *zstd.Decoder
+
+	done chan struct{}
+}
+
+func FromHost(log zerolog.Logger, host string) (*Live, error) {
+	psock, err := sub.NewSocket()
+	if err != nil {
+		return nil, fmt.Errorf("could not get new sub socket: %w", err)
+	}
+	// TODO: Switch to WSS
+	err = psock.Dial(fmt.Sprint("ws://", host, ":", pubSubPort))
+	if err != nil {
+		return nil, fmt.Errorf("could not dial on sub socket: %w", err)
+	}
+	// Empty byte array effectively subscribes to everything
+	err = psock.SetOption(mangos.OptionSubscribe, []byte("trie_updates"))
+	if err != nil {
+		return nil, fmt.Errorf("could not subscribe: %w", err)
+	}
+
+	// rsock, err := req.NewSocket()
+	// if err != nil {
+	// 	return nil, fmt.Errorf("could not get new req socket: %w", err)
+	// }
+	// err = rsock.Dial(fmt.Sprint(host, ":", reqPort))
+	// if err != nil {
+	// 	return nil, fmt.Errorf("could not dial on req socket: %w", err)
+	// }
+
+	decompressor, err := zstd.NewReader(nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize decompressor: %w", err)
+	}
+
+	live := Live{
+		log: log.With().Str("component", "live-feeder").Logger(),
+
+		pubsub:       psock,
+		decompressor: decompressor,
+
+		done: make(chan struct{}),
+	}
+
+	return &live, nil
+}
+
+func (l *Live) Run() {
+	go func() {
+		for {
+			select {
+			case <-l.done:
+				return
+			default:
+			}
+
+			msg, err := l.pubsub.Recv()
+			if err != nil {
+				l.log.Warn().Err(err).Msg("could not receive update")
+				continue
+			}
+
+			l.log.Info().Msg("recevied trie update")
+
+			// Get payload from message by removing topic header.
+			payload := bytes.Replace(msg, []byte("trie_updates"), []byte{}, 1)
+
+			payload, err = l.decompressor.DecodeAll(payload, nil)
+			if err != nil {
+				l.log.Error().Err(err).Msg("unable to decompress block payload")
+			}
+
+			var update ledger.TrieUpdate
+			err = cbor.Unmarshal(payload, &update)
+			if err != nil {
+				l.log.Error().Err(err).Msg("unable to decode block payload")
+			}
+			// TODO: Detect if we're missing some info and request it using the req.
```

Comment:
You mean FIXME? :D

But really, what info could we be missing? Either the update is complete or it isn't.

Reply:
We could have skipped an update due to a network failure, in which case we can't proceed unless we get that specific update, since the mapper is sequential and can't jump over an update